File: src/engraft/actions/json_replace.py
import json
import re
from dataclasses import dataclass
from pathlib import Path

from engraft.actions import register
from engraft.actions.base import Action


def _parse_path(dot_path: str) -> list[str | int]:
    """Parse a dot-path like 'a.b[2].c' into ['a', 'b', 2, 'c']."""
    parts: list[str | int] = []
    for segment in re.split(r"\.", dot_path):
        # Handle array indices: "items[0]" -> "items", 0
        bracket_match = re.match(r"^([^\[]*)\[(\d+)\]$", segment)
        if bracket_match:
            key, index = bracket_match.groups()
            if key:
                parts.append(key)
            parts.append(int(index))
        else:
            parts.append(segment)
    return parts


def _set_at_path(obj: dict | list, path: list[str | int], value: str) -> None:
    """Set a value at a nested path, creating intermediate dicts if needed."""
    for i, key in enumerate(path[:-1]):
        next_key = path[i + 1]
        if isinstance(key, int):
            if isinstance(obj, list) and key < len(obj):
                obj = obj[key]
            else:
                raise IndexError(f"Array index {key} out of range")
        else:
            if key not in obj:
                obj[key] = [] if isinstance(next_key, int) else {}
            obj = obj[key]

    last = path[-1]
    if isinstance(last, int):
        obj[last] = value
    else:
        obj[last] = value


@register("json_replace")
@dataclass
class JsonReplace(Action):
    """Replace a value in a JSON file using a JSONPath-like selector."""

    target: Path
    selector: str
    value: str

    def apply(self) -> None:
        data = json.loads(self.target.read_text())

        selector = self.selector
        if selector.startswith("$."):
            selector = selector[2:]

        parsed = _parse_path(selector)
        _set_at_path(data, parsed, self.value)

        self.target.write_text(json.dumps(data, indent=2) + "\n")

File: typescript/src/actions/jsonReplace.ts
import { promises as fs } from "node:fs";
import path from "node:path";

import type { Action, Variables } from "./base.js";
import { register } from "./registry.js";

type JsonValue =
  | string
  | number
  | boolean
  | null
  | JsonValue[]
  | { [k: string]: JsonValue };

type PathPart = string | number;

const parsePath = (dotPath: string): PathPart[] => {
  const parts: PathPart[] = [];
  for (const segment of dotPath.split(".")) {
    const bracket = /^([^\[]*)\[(\d+)\]$/.exec(segment);
    if (bracket) {
      const [, key, index] = bracket;
      if (key) parts.push(key);
      parts.push(Number(index));
    } else {
      parts.push(segment);
    }
  }
  return parts;
};

const setAtPath = (
  obj: JsonValue,
  keys: PathPart[],
  value: string,
): void => {
  let cursor: JsonValue = obj;
  for (let i = 0; i < keys.length - 1; i++) {
    const key = keys[i]!;
    const next = keys[i + 1]!;
    if (typeof key === "number") {
      if (!Array.isArray(cursor) || key >= cursor.length) {
        throw new Error(`Array index ${key} out of range`);
      }
      cursor = cursor[key]!;
    } else {
      if (
        cursor === null ||
        typeof cursor !== "object" ||
        Array.isArray(cursor)
      ) {
        throw new Error(`Cannot descend into non-object at key ${key}`);
      }
      const container = cursor as { [k: string]: JsonValue };
      if (!(key in container)) {
        container[key] = typeof next === "number" ? [] : {};
      }
      cursor = container[key]!;
    }
  }
  const last = keys[keys.length - 1]!;
  if (typeof last === "number") {
    if (!Array.isArray(cursor)) {
      throw new Error(`Cannot set index ${last} on non-array`);
    }
    cursor[last] = value;
  } else {
    if (
      cursor === null ||
      typeof cursor !== "object" ||
      Array.isArray(cursor)
    ) {
      throw new Error(`Cannot set key ${last} on non-object`);
    }
    (cursor as { [k: string]: JsonValue })[last] = value;
  }
};

interface ReplaceEntry {
  selector: string;
  variable: string;
}

@register("json_replace")
export class JsonReplace implements Action {
  private readonly file: string;
  private readonly replace: ReplaceEntry[];

  constructor(config: Record<string, unknown>) {
    this.file = String(config.file);
    this.replace = (config.replace as ReplaceEntry[] | undefined) ?? [];
  }

  async apply(
    variables: Variables,
    workDir: string,
    _valuesDir: string,
  ): Promise<void> {
    const target = path.join(workDir, this.file);
    const text = await fs.readFile(target, "utf8");
    const data = JSON.parse(text) as JsonValue;

    for (const entry of this.replace) {
      const value = variables[entry.variable];
      if (value == null) continue;

      let selector = entry.selector;
      if (selector.startsWith("$.")) {
        selector = selector.slice(2);
      }
      const parsed = parsePath(selector);
      setAtPath(data, parsed, value);
    }

    await fs.writeFile(target, JSON.stringify(data, null, 2) + "\n");
  }

  targetFiles(): string[] {
    return [this.file];
  }
}
